### PR TITLE
Timestamp off-chain API

### PR DESCRIPTION
**Pull request by gautamdhameja**
Implements timestamp API for off-chain workers.

### CODE REVIEW

File: core/offchain/src/api.rs [hunk 1]
```diff
@@ -188,7 +190,19 @@ impl<Storage, KeyProvider> OffchainExt for Api<Storage, KeyProvider> where
 	}
 
 	fn timestamp(&mut self) -> Timestamp {
-		unavailable_yet("timestamp")
+		let now = SystemTime::now();
+		let epoch_duration = now.duration_since(SystemTime::UNIX_EPOCH);
+		match epoch_duration {
+			Err(_) => {
+				warn!("Current time is earlier than UNIX_EPOCH.");
```
Comment (DemiMarie-parity) [inline core/offchain/src/api.rs]: It is okay to panic in this case, but as this should only affect horribly misconfigured Windows systems, it does not matter.

File: core/offchain/src/api.rs [hunk 2]
```diff
@@ -369,6 +383,22 @@ mod tests {
...
+		
+		// Get timestamp from std.
+		let now = SystemTime::now();
+		let d: u64 = now.duration_since(SystemTime::UNIX_EPOCH).unwrap().as_millis().try_into().unwrap();
+
+		// Get timestamp from offchain api.
+		let timestamp = api.timestamp();
+		
+		// Compare.
+		assert_eq!(timestamp.unix_millis() > 0, true);
```
Comment (bkchr) [inline core/offchain/src/api.rs]: Just use `assert!`

File: core/offchain/src/api.rs [hunk 3]
```diff
@@ -188,7 +190,19 @@ impl<Storage, KeyProvider> OffchainExt for Api<Storage, KeyProvider> where
...
+		let epoch_duration = now.duration_since(SystemTime::UNIX_EPOCH);
+		match epoch_duration {
+			Err(_) => {
+				warn!("Current time is earlier than UNIX_EPOCH.");
+				Timestamp::from_unix_millis(0)
+				},
+			Ok(d) => {
+				let duration = d.as_millis();
+				// Assuming overflow won't happen for a few hundred years.
+				Timestamp::from_unix_millis(duration.try_into().unwrap())
```
Comment (bkchr) [inline core/offchain/src/api.rs]: No plain unwraps in the code, please use an expect.

File: core/offchain/src/api.rs [hunk 4]
```diff
@@ -188,7 +190,19 @@ impl<Storage, KeyProvider> OffchainExt for Api<Storage, KeyProvider> where
...
 
 	fn timestamp(&mut self) -> Timestamp {
-		unavailable_yet("timestamp")
+		let now = SystemTime::now();
+		let epoch_duration = now.duration_since(SystemTime::UNIX_EPOCH);
+		match epoch_duration {
+			Err(_) => {
+				warn!("Current time is earlier than UNIX_EPOCH.");
+				Timestamp::from_unix_millis(0)
+				},
```
Comment (bkchr) [inline core/offchain/src/api.rs]: Formatting of?

File: core/offchain/src/api.rs (hunk 1 above)
Comment (tomusdrw) [reply core/offchain/src/api.rs]: I'd rather keep it as non panicker and just default to `0` without even printing a warning.

File: core/offchain/src/api.rs [hunk 5]
```diff
@@ -188,7 +190,19 @@ impl<Storage, KeyProvider> OffchainExt for Api<Storage, KeyProvider> where
...
+			Ok(d) => {
+				let duration = d.as_millis();
+				// Assuming overflow won't happen for a few hundred years.
```
Comment (tomusdrw) [inline core/offchain/src/api.rs]: It's actually billions of years :boom: This comment should end up as an `expect`

File: core/offchain/src/api.rs [hunk 6]
```diff
@@ -188,7 +190,20 @@ impl<Storage, KeyProvider> OffchainExt for Api<Storage, KeyProvider> where
...
+		match epoch_duration {
+			Err(_) => {
+				// Current time is earlier than UNIX_EPOCH.
+				Timestamp::from_unix_millis(0)
+			},
+			Ok(d) => {
+				let duration = d.as_millis();
+				// Assuming overflow won't happen for a few hundred years.
+				Timestamp::from_unix_millis(duration.try_into()
+					.expect("Overflow occurred in milliseconds conversion."))
```
Comment (bkchr) [inline core/offchain/src/api.rs]: Please follow the pattern we normally use for expect messages. (just search the codebase)

**Comment by tomusdrw** [comment]
Something bad happened during a merge, it seems to contain too many changes.

**Comment by gautamdhameja** [comment]
@tomusdrw Yeah, it seems I messed up during rebase with master. Fixing it.

**Pull request merged**